[14.0] add stock_intercompany module

ajaniszewska-dev opened the pull request.

ajaniszewska-dev committed:
[14.0] add stock_intercompany module

ajaniszewska-dev commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,26 @@
+                }
+
+                new_picking_vals = picking.copy_data(default=vals)
+                stock_move = self.env['stock.picking'].create(new_picking_vals)
```
@yvaucher  now i guess stock.picking needs to have created stock.move and stock.package.quant for partner's company and assigned to move_lines/move_ids_without package?  i think location will be taken automatically for partner on picking since it is readonly param.

i dont understand this part:

" You may have to duplicate moves 1 by 1 changing the company and it's src and destination locations." 

then we duplicate the moves or we create the moves for picking?

gurneyalex replied:
you create new moves for the new picking. I assume what Yannick means is that the easiest way is to copy the moves from the picking being processed, changing some values, as this will copy some fields that can be added by other addons that you are not aware of.

gurneyalex replied:
```suggestion
                reception_picking = self.env['stock.picking'].create(new_picking_vals)
```

ajaniszewska-dev commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,27 @@
+                    'company_id': company.id,
+                    'picking_type_id': company.intercompany_in_type_id.id  or warehouse.in_type_id.id,
+                    'state':'draft',
+                    'location_id': self.env['stock.location'].sudo().search([('name','=','Vendors')]).id,
```
@yvaucher  not sure about this location, there is Vendor location but no company_id on it.

yvaucher replied:
It's fine partner location are generic.
Would be better to do it as done in `stock` module. You can rely on the XML ids.

As done in:
https://github.com/odoo/odoo/blob/14.0/addons/stock/models/stock_picking.py#L177-L182

ajaniszewska-dev commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,27 @@
+                    'location_dest_id': self.env['stock.location'].sudo().search([('name','=','Stock'),('company_id','=',company.id)]).id,
+                }
+
+                new_picking_vals = picking.sudo().copy_data(default=vals)
```
@yvaucher  as far as i see we have always 2 counter- picking on partner company created. probably copy_data creates one? Will check it soon.

yvaucher replied:
~Maybe related to some recursive call?~

EDIT: related to the wizard when no qty are selected. -> better to override `_action_done`

yvaucher commented on stock_intercompany/__manifest__.py:
```diff
@@ -0,0 +1,17 @@
+    "Summary": "Module that adds possibility for intercompany Delivery-Reception",
+    "version": "14.0.1.1.1",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-workflow",
```
```suggestion
    "website": "https://github.com/OCA/multi-company",
```

yvaucher commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,27 @@
+
+                new_picking_vals = picking.sudo().copy_data(default=vals)
+                reception_picking = self.env['stock.picking'].sudo().create(new_picking_vals)
+                reception_picking.action_confirm()
```
Probably better to keep the reception picking in draft so the destination company user can handle in with more options.

yvaucher reviewed (changes requested):
Some remarks on the code.

Please also install `pre-commit`

yvaucher commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,27 @@
+                new_picking_vals = picking.sudo().copy_data(default=vals)
+                reception_picking = self.env['stock.picking'].sudo().create(new_picking_vals)
+                reception_picking.action_confirm()
+        return super(StockPicking,self).button_validate()
```
Would make more sense to move the `super` prior to the creation of the reception picking.

yvaucher commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,27 @@
+
+
+    #override of method from stock module
+    def button_validate(self):
```
Better to move this part in a dedicated method + override `_action_done` insead of `button_validate`

ajaniszewska-dev commented on stock_intercompany/tests/test_intercompany_picking.py:
```diff
@@ -0,0 +1,99 @@
+                "name": "Product A",
+                "type": "product",
+                "categ_id": self.env.ref("product.product_category_all").id,
+                "qty_available": 100,
```
don't understand why checking product with default_company_id inside with_context(...) shows qty_available -1.0 for both company1 and company2. with this not possible to make a  assert and check product quantities on both companies when pickings are created.

gurneyalex commented on stock_intercompany/views/res_config_settings.xml:
```diff
@@ -0,0 +1,25 @@
+                    id="intercompany_delivery_sbox"
+                >
+                    <div class="o_setting_right_pane">
+                        <label for="intercompany_in_type_id" />
```
```suggestion
                        <label for="intercompany_in_type_id" />
                        <span class="fa fa-lg fa-building-o" title="Values set here are company-specific." groups="base.group_multi_company"/>
```

this is important for UI consistency: this is set on all res.config.settings parameters which are related fields on the current company.

gurneyalex commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,33 @@
+    def create_counterpart_picking(self):
+        companies = self.env["res.company"].sudo().search([])
+        partners = {cp.partner_id: cp for cp in companies}
+        if self.partner_id in partners.keys():
```
```suggestion
        if self.partner_id in partners:
```

the "in" operator searches on keys in a dictionary

gurneyalex replied:
(!) Actually there is another issue here: I think that this code we will create a counterpart also if we purchase from another company, but the counterpart will be wrong, as it will be a reception picking on the other company (which is a supplier in this case).

For now, I suggest checking that the current picking is a shipping picking (for instance by checking that the location_dest_id of the picking is a Customers virtual location). If not, then the method should do nothing. 

THIS NEEDS FIXING.

gurneyalex commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,33 @@
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def create_counterpart_picking(self):
```
```suggestion
    def _create_counterpart_picking(self):
```

this is an internal method, which should not be exposed to RPC

gurneyalex commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,33 @@
+    # override of method from stock module
+    def _action_done(self):
+        for picking in self:
+            picking.create_counterpart_picking()
```
```suggestion
            picking._create_counterpart_picking()
```

gurneyalex commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,33 @@
+                self.env["stock.warehouse"]
+                .sudo()
+                .search([("company_id", "=", company.id)], limit=1)
+            )
```
I think it is better to use the warehouse of the picking type we configured in the res.config.settings:
```suggestion
            if `company.intercompany_in_type_id.warehouse_id:`
                warehouse = company.intercompany_in_type_id.warehouse_id
            else:
                warehouse = (
                    self.env["stock.warehouse"]
                    .sudo()
                    .search([("company_id", "=", company.id)], limit=1)
                )
```

gurneyalex commented on stock_intercompany/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,4 @@
+* Go to Inventory > Settings
+* Under 'Operations' set
+
+  * Intercompany operation type
```
add information about the creation of the picking type (which type of picking type has to be used, how to create it), on which company you need to set it.

gurneyalex committed:
[IMP] add a post action_done hook

which can be use if extension modules need to perform
some actions depending on both the picking and its counterpart

gurneyalex committed:
[IMP] documentation

leemannd committed:
14.0 Fix stock_picking: use the source partner for the new created picking

leemannd commented:
Hello @ajaniszewska-dev  I have a small fix to be pushed to your PR -> https://github.com/ajaniszewska-dev/multi-company/pull/1

leemannd commented on stock_intercompany/models/stock_picking.py:
```diff
@@ -0,0 +1,49 @@
+    def _finalize_counterpart_picking(self, counterpart_picking):
+        """hook to finalize required steps on the counterpart picking after the initial
+        outgoing picking is done"""
+        pass
```
Pre-commit failing on this point

ajaniszewska-dev committed:
Merge pull request #1 from leemannd/stock_intercompany

14.0 Fix stock_picking: use the source partner for the new created pi…

leemannd reviewed (approved):
LGTM unless the pre-commit hook failing

gurneyalex commented:
/ocabot merge nobump
